diff --git a/src/gui_jobwindow.cpp b/src/gui_jobwindow.cpp
index 2b26ff1a..a910ab28 100644
--- a/src/gui_jobwindow.cpp
+++ b/src/gui_jobwindow.cpp
@@ -1487,6 +1487,8 @@ void JobWindow::initialiseClass3DWindow()
 
     place("fn_ref", TOGGLE_DEACTIVATE);
     place("fn_mask");
+    place("fn_protein_mask");
+    place("low_pass_composite");
 
 	current_y += STEPY/2;
 
@@ -1690,6 +1692,8 @@ void JobWindow::initialiseAutorefineWindow()
 
     place("fn_ref", TOGGLE_DEACTIVATE);
     place("fn_mask");
+    place("fn_protein_mask");
+    place("low_pass_composite");
 
     current_y += STEPY/2;
 
diff --git a/src/ml_optimiser.cpp b/src/ml_optimiser.cpp
index 2a28fc90..ae4a1ee3 100644
--- a/src/ml_optimiser.cpp
+++ b/src/ml_optimiser.cpp
@@ -229,8 +229,12 @@ void MlOptimiser::parseContinue(int argc, char **argv)
     if (fnt != "OLD")
         fn_mask2 = fnt;
 
+    // write_every_iter
+    write_every_iter = textToInteger(parser.getOption("--write_iter", "Write out model every so many iterations (default is writing out all iters)", "1"));
+
     // These are still experimental; so not in the optimiser.star yet.
     fn_lowpass_mask = parser.getOption("--lowpass_mask", "User-provided mask for low-pass filtering", "None");
+    fn_lowpass_mask_composite = parser.getOption("--lowpass_mask_composite", "User-provided mask for composite low-pass filtering", "None");
     lowpass = textToFloat(parser.getOption("--lowpass", "User-provided cutoff for region specified above", "0"));
 
     // Check whether tau2-spectrum has changed
@@ -357,6 +361,10 @@ void MlOptimiser::parseContinue(int argc, char **argv)
         }
     }
 
+    //Check for relax_symmetry option
+    std::string sym_relax_ =  parser.getOption("--relax_sym", "Symmetry to be relaxed", "");
+    sampling.fn_sym_relax = sym_relax_;
+
     fnt = parser.getOption("--auto_local_healpix_order", "Minimum healpix order (before oversampling) from which auto-refine procedure will use local searches", "OLD");
     if (fnt != "OLD")
         autosampling_hporder_local_searches = textToInteger(fnt);
@@ -556,6 +564,8 @@ void MlOptimiser::parseContinue(int argc, char **argv)
 
     do_trust_ref_size = parser.checkOption("--trust_ref_size", "Trust the pixel and box size of the input reference; by default the program will die if these are different from the first optics group of the data");
 
+    fast_subsets_min_parts_per_class = textToInteger(parser.getOption("--fast_subsets_min_parts_per_class", "Minimal particles per class when using faster optimisation", "0"));
+
     // Debugging/analysis/hidden stuff
     do_map = !checkParameter(argc, argv, "--no_map");
     minres_map = textToInteger(getParameter(argc, argv, "--minres_map", "5"));
@@ -643,11 +653,13 @@ void MlOptimiser::parseInitial(int argc, char **argv)
     fn_mask = parser.getOption("--solvent_mask", "User-provided mask for the references (default is to use spherical mask with particle_diameter)", "None");
     fn_mask2 = parser.getOption("--solvent_mask2", "User-provided secondary mask (with its own average density)", "None");
     fn_lowpass_mask = parser.getOption("--lowpass_mask", "User-provided mask for low-pass filtering", "None");
+    fn_lowpass_mask_composite = parser.getOption("--lowpass_mask_composite", "User-provided mask for composite low-pass filtering", "None");
     lowpass = textToFloat(parser.getOption("--lowpass", "User-provided cutoff for region specified above", "0"));
     fn_tau = parser.getOption("--tau", "STAR file with input tau2-spectrum (to be kept constant)", "None");
     fn_local_symmetry = parser.getOption("--local_symmetry", "Local symmetry description file containing list of masks and their operators", "None");
     do_split_random_halves = parser.checkOption("--split_random_halves", "Refine two random halves of the data completely separately");
     low_resol_join_halves = textToFloat(parser.getOption("--low_resol_join_halves", "Resolution (in Angstrom) up to which the two random half-reconstructions will not be independent to prevent diverging orientations","-1"));
+    write_every_iter = textToInteger(parser.getOption("--write_iter", "Write out model every so many iterations (default is writing out all iters)", "1"));
     do_center_classes = parser.checkOption("--center_classes", "Re-center classes based on their center-of-mass?");
 
     // Initialisation
@@ -903,6 +915,7 @@ void MlOptimiser::parseInitial(int argc, char **argv)
     do_reuse_scratch = parser.checkOption("--reuse_scratch", "Re-use data on scratchdir, instead of wiping it and re-copying all data.");
     keep_scratch = parser.checkOption("--keep_scratch", "Don't remove scratch after convergence. Following jobs that use EXACTLY the same particles should use --reuse_scratch.");
     do_fast_subsets = parser.checkOption("--fast_subsets", "Use faster optimisation by using subsets of the data in the first 15 iterations");
+    fast_subsets_min_parts_per_class = textToInteger(parser.getOption("--fast_subsets_min_parts_per_class", "Minimal particles per class when using faster optimisation", "0"));
 #ifdef ALTCPU
     do_cpu = parser.checkOption("--cpu", "Use intel vectorisation implementation for CPU");
 #else
@@ -1352,7 +1365,8 @@ void MlOptimiser::read(FileName fn_in, int rank, bool do_prevent_preread)
 
 void MlOptimiser::write(bool do_write_sampling, bool do_write_data, bool do_write_optimiser, bool do_write_model, int random_subset)
 {
-    if (do_grad && subset_size > 0 && (iter % write_every_grad_iter) != 0 && iter != nr_iter)
+//    if (do_grad && subset_size > 0 && (iter % write_every_grad_iter) != 0 && iter != nr_iter)
+    if (((do_grad && subset_size > 0 && (iter % write_every_grad_iter) != 0) || (iter > -1 && (iter % write_every_iter) != 0)) && iter != nr_iter)
         return;
 
     FileName fn_root, fn_tmp, fn_model, fn_model2, fn_data, fn_sampling, fn_root2;
@@ -2489,6 +2503,7 @@ void MlOptimiser::initialiseGeneral(int rank)
     // Check mask angpix, boxsize and [0,1] compliance right away.
     if (fn_mask != "None") checkMask(fn_mask, 1, rank);
     if (fn_mask2 != "None") checkMask(fn_mask2, 2, rank);
+    if (fn_lowpass_mask_composite != "None") checkMask(fn_lowpass_mask_composite, 3, rank);
 
     // Write out unmasked 2D class averages
     do_write_unmasked_refs = (mymodel.ref_dim == 2 && !gradient_refine);
@@ -5471,7 +5486,7 @@ void MlOptimiser::solventFlatten()
         return;
 
     // First read solvent mask from disc, or pre-calculate it
-    Image<RFLOAT> Isolvent, Isolvent2, Ilowpass;
+    Image<RFLOAT> Isolvent, Isolvent2, Ilowpass, Ilowpass_composite;
     Isolvent().resize(mymodel.Iref[0]);
     Isolvent().setXmippOrigin();
     Isolvent().initZeros();
@@ -5541,9 +5556,18 @@ void MlOptimiser::solventFlatten()
             REPORT_ERROR("MlOptimiser::solventFlatten ERROR: second solvent mask is of incorrect size.");
     }
 
+    // Also read a composite lowpass mask if necessary
+    if (fn_lowpass_mask_composite != "None")
+    {
+        Ilowpass_composite.read(fn_lowpass_mask_composite);
+        Ilowpass_composite().setXmippOrigin();
+        if (!Ilowpass_composite().sameShape(Isolvent()))
+            REPORT_ERROR("MlOptimiser::solventFlatten ERROR: composite lowpass mask is of incorrect size.");
+    }
+
     for (int iclass = 0; iclass < mymodel.nr_classes; iclass++)
     {
-        MultidimArray<RFLOAT> Itmp;
+        MultidimArray<RFLOAT> Itmp, Itmp_composite;
         if (fn_lowpass_mask != "None")
         {
             Itmp = mymodel.Iref[iclass];
@@ -5551,8 +5575,23 @@ void MlOptimiser::solventFlatten()
             lowPassFilterMap(Itmp, lowpass, mymodel.pixel_size);
         }
 
-        // Then apply the expanded solvent mask to the map
-        mymodel.Iref[iclass] *= Isolvent(); // this is the tight mask
+        if (fn_lowpass_mask_composite != "None" && lowpass > 0.)
+        {
+            Itmp_composite = mymodel.Iref[iclass];
+            lowPassFilterMap(Itmp_composite, lowpass, mymodel.pixel_size);
+            Itmp_composite *= Ilowpass_composite();
+            RFLOAT solv;
+            FOR_ALL_DIRECT_ELEMENTS_IN_ARRAY3D(Isolvent())
+            {
+                solv = DIRECT_A3D_ELEM(Isolvent(), k, i, j);
+                DIRECT_A3D_ELEM(mymodel.Iref[iclass], k, i, j) = solv * DIRECT_A3D_ELEM(mymodel.Iref[iclass], k, i, j) + (1. - solv) * DIRECT_A3D_ELEM(Itmp_composite, k, i, j);
+            }
+        }
+        else
+        {
+            // Then apply the expanded solvent mask to the map
+            mymodel.Iref[iclass] *= Isolvent(); // this is the tight mask
+        }
 
         if (fn_lowpass_mask != "None")
             mymodel.Iref[iclass] += Itmp;
@@ -9963,6 +10002,7 @@ void MlOptimiser::updateSubsetSize(bool myverb)
     if (do_fast_subsets)
     {
         long int min_parts_per_class = (mymodel.ref_dim == 2) ? 100 : 1500;
+	if (fast_subsets_min_parts_per_class > 0) min_parts_per_class = fast_subsets_min_parts_per_class;
         if (iter <= 5)
         {
             subset_size = min_parts_per_class*mymodel.nr_classes;
diff --git a/src/ml_optimiser.h b/src/ml_optimiser.h
index 3004cded..f236b2cc 100644
--- a/src/ml_optimiser.h
+++ b/src/ml_optimiser.h
@@ -343,8 +343,12 @@ public:
 
 	// Mask for regions to be low-pass filtered
 	FileName fn_lowpass_mask;
+	FileName fn_lowpass_mask_composite;
 	RFLOAT lowpass;
 
+	// write out model very X iterations.
+	int write_every_iter;
+
 	// Width of the soft-edges of the circular masks
 	int width_mask_edge;
 
@@ -447,6 +451,7 @@ public:
 
 	// Use subsets like in cisTEM to speed up 2D/3D classification
 	bool do_fast_subsets;
+	long int fast_subsets_min_parts_per_class;
 
 	// Available memory (in Gigabyte)
 	size_t available_gpu_memory;
diff --git a/src/pipeline_jobs.cpp b/src/pipeline_jobs.cpp
index 9d0e3044..a9bb86fe 100644
--- a/src/pipeline_jobs.cpp
+++ b/src/pipeline_jobs.cpp
@@ -3644,6 +3644,26 @@ Thereby, for example, the higher density inside the virion may be set to a const
 Note that this second mask should have one-values inside the virion and zero-values in the capsid and the solvent areas. \
 To use a second mask, use the additional option --solvent_mask2, which may given in the Additional arguments line (in the Running tab).");
 
+	joboptions["fn_protein_mask"] = JobOption("Mask of protein of interest (optional):", LABEL_MASK_CPIPE, 1, "", "Image Files (*.{spi,vol,msk,mrc})", "\
+Tight (soft) mask of the protein of interest.\n\
+\n\
+When this protein mask and the composite low-pass filter are provided, the voxels outside the protein mask and within the reference mask will be low-pass filtered. \
+This is the composite masking method used in:\n\
+\n\
+Dou, T., Lian, T., Shu, S. et al. The substrate and inhibitor binding mechanism of polyspecific transporter OAT1 revealed by high-resolution cryo-EM. Nat Struct Mol Biol 30, 1794–1805 (2023). https://doi.org/10.1038/s41594-023-01123-3.\n\
+\n\
+and\n\
+\n\
+Jiang, Y., Jiang, J. The Bor1 elevator transport cycle is subject to autoinhibition and activation. Nat Commun 15, 9090 (2024). https://doi.org/10.1038/s41467-024-53411-1\n\
+\n\
+To use composite masks for membrane proteins, provide a regular reference mask, that covers the entire structure of both the protein(s) and the detergent micelle, \
+and a soft mask of protein of interest excluding the detergent micelle, and set a low-pass filter resolution.\n\
+");
+	joboptions["low_pass_composite"] = JobOption("Composite low-pass filter (A) (optional):", -1, 1, 50, 1, "\
+Used together with the mask of protein of interest for composite masking. \
+Usually, 20 A is good to eliminate over-fitting of the detergent/lipid micelles of membrane proteins.\
+");
+
 	joboptions["ref_correct_greyscale"] = JobOption("Ref. map is on absolute greyscale?", false, "Probabilities are calculated based on a Gaussian noise model, \
 which contains a squared difference term between the reference and the experimental image. This has a consequence that the \
 reference needs to be on the same absolute intensity grey-scale as the experimental images. \
@@ -3976,7 +3996,20 @@ bool RelionJob::getCommandsClass3DJob(std::string &outputname, std::vector<std::
 
 	if (joboptions["fn_mask"].getString().length() > 0)
 	{
-		command += " --solvent_mask " + joboptions["fn_mask"].getString();
+		if (joboptions["fn_protein_mask"].getString().length() > 0 && joboptions["low_pass_composite"].getNumber(error_message) > 0)
+		{
+			// use the protein mask as the solvent mask.
+			command += " --solvent_mask " + joboptions["fn_protein_mask"].getString();
+			command += " --lowpass_mask_composite " + joboptions["fn_mask"].getString();
+			command += " --lowpass " + joboptions["low_pass_composite"].getString();
+			Node node(joboptions["fn_protein_mask"].getString(), joboptions["fn_protein_mask"].node_type);
+			inputNodes.push_back(node);
+		}
+		else
+		{
+			command += " --solvent_mask " + joboptions["fn_mask"].getString();
+		}
+
 		Node node(joboptions["fn_mask"].getString(), joboptions["fn_mask"].node_type);
 		inputNodes.push_back(node);
 	}
@@ -4158,6 +4191,26 @@ Thereby, for example, the higher density inside the virion may be set to a const
 Note that this second mask should have one-values inside the virion and zero-values in the capsid and the solvent areas. \
 To use a second mask, use the additional option --solvent_mask2, which may given in the Additional arguments line (in the Running tab).");
 
+	joboptions["fn_protein_mask"] = JobOption("Mask of protein of interest (optional):", LABEL_MASK_CPIPE, 1, "", "Image Files (*.{spi,vol,msk,mrc})", "\
+Tight (soft) mask of the protein of interest.\n\
+\n\
+When this protein mask and the composite low-pass filter are provided, the voxels outside the protein mask and within the reference mask will be low-pass filtered. \
+This is the composite masking method used in:\n\
+\n\
+Dou, T., Lian, T., Shu, S. et al. The substrate and inhibitor binding mechanism of polyspecific transporter OAT1 revealed by high-resolution cryo-EM. Nat Struct Mol Biol 30, 1794–1805 (2023). https://doi.org/10.1038/s41594-023-01123-3.\n\
+\n\
+and\n\
+\n\
+Jiang, Y., Jiang, J. The Bor1 elevator transport cycle is subject to autoinhibition and activation. Nat Commun 15, 9090 (2024). https://doi.org/10.1038/s41467-024-53411-1\n\
+\n\
+To use composite masks for membrane proteins, provide a regular reference mask, that covers the entire structure of both the protein(s) and the detergent micelle, \
+and a soft mask of protein of interest excluding the detergent micelle, and set a low-pass filter resolution.\n\
+");
+	joboptions["low_pass_composite"] = JobOption("Composite low-pass filter (A) (optional):", -1, 1, 50, 1, "\
+Used together with the mask of protein of interest for composite masking. \
+Usually, 20 A is good to eliminate over-fitting of the detergent/lipid micelles of membrane proteins.\
+");
+
 	joboptions["ref_correct_greyscale"] = JobOption("Ref. map is on absolute greyscale?", false, "Probabilities are calculated based on a Gaussian noise model, \
 which contains a squared difference term between the reference and the experimental image. This has a consequence that the \
 reference needs to be on the same absolute intensity grey-scale as the experimental images. \
@@ -4464,7 +4517,19 @@ bool RelionJob::getCommandsAutorefineJob(std::string &outputname, std::vector<st
 	}
 	if (joboptions["fn_mask"].getString().length() > 0)
 	{
-		command += " --solvent_mask " + joboptions["fn_mask"].getString();
+		if (joboptions["fn_protein_mask"].getString().length() > 0 && joboptions["low_pass_composite"].getNumber(error_message) > 0)
+		{
+			// use the protein mask as the solvent mask.
+			command += " --solvent_mask " + joboptions["fn_protein_mask"].getString();
+			command += " --lowpass_mask_composite " + joboptions["fn_mask"].getString();
+			command += " --lowpass " + joboptions["low_pass_composite"].getString();
+			Node node(joboptions["fn_protein_mask"].getString(), joboptions["fn_protein_mask"].node_type);
+			inputNodes.push_back(node);
+		}
+		else
+		{
+			command += " --solvent_mask " + joboptions["fn_mask"].getString();
+		}
 
 		if (joboptions["do_solvent_fsc"].getBoolean())
 			command += " --solvent_correct_fsc ";
diff --git a/src/postprocessing.cpp b/src/postprocessing.cpp
index e2ee9d14..5a9d80e9 100644
--- a/src/postprocessing.cpp
+++ b/src/postprocessing.cpp
@@ -29,6 +29,7 @@ void Postprocessing::read(int argc, char **argv)
 	fn_OS = parser.getOption("--ios", "Input tomo optimiser set file. It is used to set --i if not provided. Updated output optimiser set is created.", "");
 	fn_out = parser.getOption("--o", "Output rootname", "postprocess");
 	angpix = textToFloat(parser.getOption("--angpix", "Pixel size in Angstroms", "-1"));
+	requested_angpix = textToFloat(parser.getOption("--rescale_angpix", "Scale output image(s) to this new pixel size (in A)", "-1."));
 	write_halfmaps = parser.checkOption("--half_maps", "Write post-processed half maps for validation");
 	mtf_angpix  = textToFloat(parser.getOption("--mtf_angpix", "Pixel size in the original micrographs/movies (in Angstroms)", "-1."));
 	molweight = textToFloat(parser.getOption("--molweight", "Molecular weight (in kDa) of ordered protein mass", "-1"));
@@ -556,6 +557,20 @@ RFLOAT Postprocessing::sharpenMap()
 		lowPassFilterMap(FT, XSIZE(I1()), applied_filter, angpix, filter_edge_width);
 	}
 
+        if (requested_angpix > 0.)
+        {
+                int oldsize = XSIZE(I1());
+                int newsize = ROUND(oldsize * (angpix / requested_angpix));
+                newsize -= newsize % 2; //make even in case it is not already
+                RFLOAT real_angpix = oldsize * angpix / newsize;
+                if (fabs(real_angpix - requested_angpix) / requested_angpix > 0.001)
+                        std::cerr << "WARNING: Although the requested pixel size (--rescale_angpix) is " << requested_angpix << " A/px, the actual pixel size will be " << real_angpix << " A/px due to rounding of the box size to an even number. The latter value is set to the image header. You can overwrite the header pixel size by --force_header_angpix." << std::endl;
+                MultidimArray<Complex > FT2;
+                windowFourierTransform(FT, FT2, newsize);
+                I1_rescaled().resize(newsize, newsize, newsize);
+                transformer.inverseFourierTransform(FT2, I1_rescaled());
+        }
+
 	transformer.inverseFourierTransform(FT, I1());
 
 	return applied_filter;
@@ -938,6 +953,17 @@ void Postprocessing::writeMaps(FileName fn_root) {
 			std::cout.width(35); std::cout << std::left   <<"  + Processed masked map: "; std::cout << fn_tmp<< std::endl;
 		}
 	}
+
+        // write rescaled map
+        if (requested_angpix > 0.)
+        {
+                I1_rescaled.setStatisticsInHeader();
+                I1_rescaled.setSamplingRateInHeader(requested_angpix);
+                fn_tmp = fn_root + "_rescaled.mrc";
+                I1_rescaled.write(fn_tmp);
+                std::cout.width(35); std::cout << std::left   <<"  + Rescaled processed map: "; std::cout << fn_tmp<< std::endl;
+        }
+
 }
 
 void Postprocessing::writeFscXml(MetaDataTable &MDfsc)
diff --git a/src/postprocessing.h b/src/postprocessing.h
index f3cbd90e..6b2caad0 100644
--- a/src/postprocessing.h
+++ b/src/postprocessing.h
@@ -50,10 +50,10 @@ public:
 	MetaDataTable optimisationSet;
 
 	// Images for the two half-reconstructions and the mask
-	Image<RFLOAT> I1, I2, Im;
+	Image<RFLOAT> I1, I2, Im, I1_rescaled;
 
 	// Pixel size in Angstroms
-	RFLOAT angpix, mtf_angpix;
+	RFLOAT angpix, mtf_angpix, requested_angpix;
 
 	/////// Masking
 
